File: clients/desktop/src/lib/game/event-bus.ts
// ══════════════════════════════════════════════
// Event Bus — 移植自 macOS EventBus.swift
// ══════════════════════════════════════════════

export type GameEventType =
  // Perception
  | 'screenCaptured' | 'windowChanged' | 'userIdle' | 'userActive'
  // Cognition
  | 'contextAnalyzed'
  // Game
  | 'questTriggered' | 'questCompleted' | 'questFailed'
  | 'buffActivated' | 'buffExpired' | 'debuffActivated' | 'debuffExpired'
  | 'levelUp' | 'expGained' | 'statChanged'
  // System
  | 'notificationPush' | 'systemStart' | 'systemStop'

export interface GameEvent {
  readonly type: GameEventType
  readonly data: Record<string, unknown>
  readonly timestamp: string
  readonly source: string
}

function createEvent(
  type: GameEventType,
  data: Record<string, unknown> = {},
  source = 'system',
): GameEvent {
  return { type, data, timestamp: new Date().toISOString(), source }
}

export type GameEventHandler = (event: GameEvent) => void

export class GameEventBus {
  private handlers = new Map<GameEventType, Map<string, GameEventHandler>>()
  private history: GameEvent[] = []
  private readonly maxHistory = 500

  on(eventType: GameEventType, handler: GameEventHandler): string {
    const id = crypto.randomUUID()
    if (!this.handlers.has(eventType)) {
      this.handlers.set(eventType, new Map())
    }
    this.handlers.get(eventType)!.set(id, handler)
    return id
  }

  off(eventType: GameEventType, id: string): void {
    this.handlers.get(eventType)?.delete(id)
  }

  emit(event: GameEvent): void
  emit(type: GameEventType, data?: Record<string, unknown>, source?: string): void
  emit(
    eventOrType: GameEvent | GameEventType,
    data?: Record<string, unknown>,
    source?: string,
  ): void {
    const event: GameEvent =
      typeof eventOrType === 'string'
        ? createEvent(eventOrType, data, source)
        : eventOrType

    this.history.push(event)
    if (this.history.length > this.maxHistory) {
      this.history.splice(0, this.history.length - this.maxHistory)
    }

    const eventHandlers = this.handlers.get(event.type)
    if (eventHandlers) {
      for (const handler of eventHandlers.values()) {
        handler(event)
      }
    }
  }

  getHistory(type?: GameEventType, limit = 50): GameEvent[] {
    const filtered = type
      ? this.history.filter((e) => e.type === type)
      : this.history
    return filtered.slice(-limit)
  }
}

File: clients/mac_old/SoloAgent/Sources/GameEngine/EventBus.swift
import Foundation

// MARK: - Event Types

enum GameEventType: String, CaseIterable {
    // Perception
    case screenCaptured
    case windowChanged
    case userIdle
    case userActive

    // Cognition
    case contextAnalyzed

    // Game
    case questTriggered
    case questCompleted
    case questFailed
    case buffActivated
    case buffExpired
    case debuffActivated
    case debuffExpired
    case levelUp
    case expGained
    case statChanged

    // System
    case notificationPush
    case systemStart
    case systemStop
}

// MARK: - Event

struct GameEvent {
    let type: GameEventType
    var data: [String: Any]
    let timestamp: Date
    let source: String

    init(type: GameEventType, data: [String: Any] = [:], source: String = "system") {
        self.type = type
        self.data = data
        self.timestamp = Date()
        self.source = source
    }
}

// MARK: - Event Bus

typealias GameEventHandler = @MainActor (GameEvent) -> Void

@MainActor
final class GameEventBus {
    private var handlers: [GameEventType: [(id: UUID, handler: GameEventHandler)]] = [:]
    private var history: [GameEvent] = []
    private let maxHistory = 500

    @discardableResult
    func on(_ eventType: GameEventType, handler: @escaping GameEventHandler) -> UUID {
        let id = UUID()
        handlers[eventType, default: []].append((id: id, handler: handler))
        return id
    }

    func off(_ eventType: GameEventType, id: UUID) {
        handlers[eventType]?.removeAll { $0.id == id }
    }

    func emit(_ event: GameEvent) {
        history.append(event)
        if history.count > maxHistory {
            history.removeFirst(history.count - maxHistory)
        }

        let eventHandlers = handlers[event.type] ?? []
        for (_, handler) in eventHandlers {
            handler(event)
        }
    }

    func emit(_ type: GameEventType, data: [String: Any] = [:], source: String = "system") {
        emit(GameEvent(type: type, data: data, source: source))
    }

    func getHistory(type: GameEventType? = nil, limit: Int = 50) -> [GameEvent] {
        let filtered: [GameEvent]
        if let type {
            filtered = history.filter { $0.type == type }
        } else {
            filtered = history
        }
        return Array(filtered.suffix(limit))
    }
}
